local-volume/provisioner/pkg/deleter/jobcontroller.go

```diff
@@ -0,0 +1,354 @@
+	} else {
+		pvName := getPVNameFromJob(job.Name)
+		// Order here is important. Delete the finished job first and then the PV.
+		glog.Infof("Job %s has completed, deleting the job", key)
+		if err := c.RuntimeConfig.APIUtil.DeleteJob(job.Name, c.namespace); err != nil {
+			if !errors.IsNotFound(err) {
+				return fmt.Errorf("Error deleting Job %q: %v", job.Name, err.Error())
+			}
+			// If the job has been deleted, we assume that the PV has also been recycled by the same processItem()
+			// call that deleted the job. There is a chance that the thread failed to delete the PV after deleting
+			// the job. In that case, a new cleaning job will eventually be launched and the whole cleaning cycle
+			// will be repeated. This is an acceptable inefficiency for this failure case, as making sure that
+			// the deletion of job and pv is done by the same processItem() call and in that fixed sequence,
+			// avoids a whole bunch of unwanted race conditions.
+			glog.Infof("Job %s has already been deleted.", key)
+			return nil
+		}
+
+		glog.Infof("Job %s has completed and been deleted. Deleting PV %s", key, pvName)
+		if err := c.APIUtil.DeletePV(pvName); err != nil {
+			if !errors.IsNotFound(err) {
+				pv, exists := c.RuntimeConfig.Cache.GetPV(pvName)
+				if exists {
+					c.RuntimeConfig.Recorder.Eventf(pv, apiv1.EventTypeWarning, common.EventVolumeFailedDelete,
+						err.Error())
+				}
+				return fmt.Errorf("Error deleting PV %q: %v", pvName, err.Error())
+			}
+		}
+	}
+
+	glog.Infof("Lister provided Job : %s", job.Name)
+	return nil
+}
+
+// IsCleaningJobRunning returns true if a cleaning job is running for the specified PV.
+func (c *jobController) IsCleaningJobRunning(pvName string) bool {
+	jobName := generateCleaningJobName(pvName)
+	_, err := c.jobLister.Jobs(c.namespace).Get(jobName)
+	if errors.IsNotFound(err) {
+		return false
+	}
+
+	if err != nil {
+		glog.Warningf("Failed to check whether job %s is running. Assuming its still running.", jobName)
+	}
+
+	// As long as the job still exists, we consider the job as running.
+	return true
+}
+
+// NewCleanupJob creates manifest for a cleaning job.
+func NewCleanupJob(pv *apiv1.PersistentVolume, imageName string, nodeName string, namespace string, command []string,
+	blkdevPath string, discoveryMap map[string]common.MountConfig) *batch_v1.Job {
+
+	// Exec scripts expect devicepath as environment variables.
+	env := []apiv1.EnvVar{{Name: common.LocalPVEnv, Value: blkdevPath}}
+	priv := true
+	// Container definition
+	jobContainer := apiv1.Container{
+		Name:    JobContainerName,
+		Image:   imageName,
+		Command: command,
+		Env:     env,
+		SecurityContext: &apiv1.SecurityContext{
+			Privileged: &priv,
+		},
+	}
+	var volumes []apiv1.Volume
+	if len(discoveryMap) > 0 {
+		jobContainer.VolumeMounts = make([]apiv1.VolumeMount, len(discoveryMap))
+		volumes = make([]apiv1.Volume, len(discoveryMap))
+		i := 0
+		for _, mount := range discoveryMap {
```

Comment:
Does it need to mount everything?  Or can it only mount the single device that needs to be cleaned?